File: src/test/stack/java/StackTest.java
package stack;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

public class StackTest {
    @Test
    public void test_isEmpty_isFull() {
        Stack stack = new Stack(3);
        
        // Verifica se a pilha está vazia no início
        assertTrue(stack.isEmpty());
        assertFalse(stack.isFull());

        // Adiciona elementos
        stack.push(1);
        stack.push(2);
        stack.push(3);

        // Tenta adicionar além do limite (deve gerar uma exceção)
        try {
            stack.push(4);
        } catch (Exception e) {
            System.out.println(e);
        }

        // Agora a pilha deve estar cheia
        assertTrue(stack.isFull());
        assertFalse(stack.isEmpty());

        // Verifica se a pilha contém os elementos corretos
        assertEquals("1, 2, 3", stack.toString()); // Removido o "4" que não deveria estar
    }

    @Test
    public void test_push_pop_toString() {
        Stack stack = new Stack(4);

        // Tenta remover de uma pilha vazia (deve gerar uma exceção)
        try {
            stack.pop();
        } catch (Exception e) {
            System.out.println(e);
        }

        // Confere se a pilha ainda está vazia
        assertTrue(stack.isEmpty());
        assertFalse(stack.isFull());

        // Adiciona elementos
        stack.push(1);
        stack.push(2);
        stack.push(3);

        // Verifica se a conversão para string está correta
        assertEquals("1, 2, 3", stack.toString());

        // Remove elementos e verifica
        stack.pop();
        stack.pop();
        assertEquals("1", stack.toString());

        stack.pop();
        assertEquals("", stack.toString()); // Pilha vazia
    }

    @Test
    public void test_peek() {
        Stack stack = new Stack(5);

        // Testa operações de `peek()` em pilha vazia
        try {
            stack.peek();
        } catch (Exception e) {
            System.out.println(e);
        }

        try {
            stack.peekLast();
        } catch (Exception e) {
            System.out.println(e);
        }

        try {
            stack.peekIndex(0);
        } catch (Exception e) {
            System.out.println(e);
        }

        // Adiciona elementos e testa `peek()`
        stack.push(3);
        stack.push(4);
        stack.push(5);
        assertEquals(5, stack.peek()); // `peek()` deve retornar o topo da pilha

        stack.pop();
        assertEquals(4, stack.peek());

        stack.push(5);
        assertEquals(3, stack.peekLast()); // Último elemento adicionado

        stack.push(6);
        assertEquals(5, stack.peekIndex(2));
        assertEquals(6, stack.peekIndex(3));

        // Testa índices inválidos
        try {
            stack.peekIndex(-1);
        } catch (Exception e) {
            System.out.println(e);
        }

        try {
            stack.peekIndex(5);
        } catch (Exception e) {
            System.out.println(e);
        }
    }
    
}

File: src/test/stack/python/stack_test.py
import unittest
import sys
import os
project_root = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))

sys.path.insert(0, os.path.join(project_root, "src/main/stack/python"))

from Stack import Stack

class TestMathOperations(unittest.TestCase):
    def test_isEmpty_isFull(self):
        stack1 = Stack(10)
        self.assertTrue(stack1.isEmpty())
        self.assertFalse(stack1.isFull())
        stack1.push(2)
        self.assertFalse(stack1.isEmpty())
        self.assertFalse(stack1.isFull())
        stack1.pop()
        self.assertTrue(stack1.isEmpty())
        stack1.push(0)
        stack1.push(10)
        stack1.push(20)
        stack1.push(30)
        stack1.push(40)
        stack1.push(50)
        stack1.push(60)
        stack1.push(70)
        stack1.push(80)
        stack1.push(100)
        self.assertEqual(stack1.toString(), "0, 10, 20, 30, 40, 50, 60, 70, 80, 100")
        try:
            stack1.push(90)
        except Exception as e:
            print("Está cheia!")
        self.assertTrue(stack1.isFull())
    
    def test_pop_push(self):
        stacks1 = Stack(5)
        self.assertTrue(stacks1.isEmpty())
        try:
            stacks1.pop()
        except Exception as e:
            print("Não ah o que popar")
        stacks1.push(2)
        stacks1.push(10)
        stacks1.pop()
        self.assertEqual(stacks1.toString(), "2")
        stacks1.push(123)
        stacks1.push(421)
        stacks1.push(53)
        stacks1.push(56)
        try:
            stacks1.push(1)
        except Exception as e:
            print("Deu erro")
        stacks1.pop()
        stacks1.pop()
        stacks1.pop()
        self.assertEqual(stacks1.toString(), "2, 123")
    
    def test_pushFirst_RemoveLast(self):
        stacks = Stack(3)
        stacks.push(3)
        stacks.push(2)
        stacks.pushFirst(1)
        self.assertEqual(stacks.toString(), "1, 3, 2")
        stacks.removeLast();
        self.assertEqual(stacks.toString(), "3, 2")
        stacks.pushFirst(1)
        try:
            stacks.pushFirst(0)
        except Exception as e:
            print("Erro")
        stacks.pop()
        stacks.pop()
        stacks.pop()
        try:
            stacks.removeLast();
        except Exception as e:
            print("erro")
    
    def test_pushIndex_peek_indexOf(self):
        stacks4 = Stack(6)
        stacks4.pushIndex(1,0)
        self.assertEqual(stacks4.toString(), "1")
        stacks4.pushIndex(2,0)
        self.assertEqual(stacks4.toString(), "2, 1")
        try:
            stacks4.pushIndex(3,1)
        except Exception as e:
            print("erro")
        stacks4.push(1)
        stacks4.push(3)
        self.assertEqual(stacks4.peekIndex(0), 2)
        self.assertEqual(stacks4.peekIndex(1), 3)
        try:
            stacks4.peekIndex(3)
        except Exception as e:
            print("erro")
        self.assertEqual(stacks4.indexOf(1), 2)
        self.assertNotEqual(stacks4.indexOf(3), 3)
        self.assertEqual(stacks4.indexOf(4), -1)
    def test_peek_peekLast(self):
        stacks = Stack(3)
        self.assertTrue(stacks.isEmpty())
        stacks.push(1)
        stacks.push(2)
        self.assertEqual(stacks.peek(),2)
        stacks.pop()
        self.assertEqual(stacks.peek(),1)
        self.assertEqual(stacks.peek(), stacks.peekLast())
        stacks.push(3)
        self.assertNotEqual(stacks.peek(), stacks.peekLast())
        self.assertEqual(stacks.peekLast(), 1)

    def test_RemoveIndex_PushINdex(self):
        stack3 = Stack(4)
        try:
            stack3.removeIndex(3)
        except Exception as e:
            print("NÃO DEU")
        stack3.pushIndex(1,0)
        stack3.pushIndex(2,1)
        #try:
        #    stack3.PushIndex(4,3)
        #except Exception as e:
        #    print("Não deu")
        stack3.pushIndex(3,2)
        self.assertEqual(stack3.toString(), "1, 2, 3")
        try:
            stack.PushIndex(3,-1)
        except Exception as e:
            print("NÃO DEU")
        stack3.removeIndex(2)
        self.assertEqual(stack3.toString(), "1, 2")


if __name__ == "__main__":
    unittest.main()
